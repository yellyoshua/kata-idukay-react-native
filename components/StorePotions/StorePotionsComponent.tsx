import React from 'react';
import { StyleSheet, TouchableOpacity, View, Text, Image } from 'react-native';
import useGame from "../../hooks/useGame"
import { PotionProps, HechiceroProps, GameStatsType } from '../../types';

export default function StorePotions({ player, resetPlayer }: { player: HechiceroProps; resetPlayer: (val: null) => void }) {
  const [potions, setPotion] = React.useState<PotionProps[]>([]);
  const [gameStats, setGameStats] = React.useState<GameStatsType | null>(null);
  const { gameResult, resetGame } = useGame();

  const addPotion = (potion: PotionProps) => {
    setPotion([...potions, potion]);
  };

  const startFigth = () => {
    const gameStats = gameResult(potions);
    setPotion([]);
    setGameStats(gameStats);
  }

  const endFigth = () => {
    resetGame(() => {
      setGameStats(null);
      resetPlayer(null);
    });
  }

  if (gameStats) {
    return <View style={styles.container}>
      <Text style={styles.title}>GAME STATS</Text>

      <View style={styles.gameStats}>
        {gameStats.combinations.map((combination, index) => {

          const isPlural = Boolean(combination > 1);

          return (
            <Text style={styles.gameStatsAttacks} key={index}>
              Ataque {index + 1}: usar
              {isPlural ? ` ${combination} ` : " una "}
              {isPlural ? " pociones " : " poción "}
              {isPlural && " distintas "}
              {isPlural ? " causan " : " causa "} un {gameStats.potionsDamage[index]}% de daño.
            </Text>
          )
        })}
        <Text style={styles.gameStatsTotal}>Total: {player && player.name} ha causado un {gameStats.damageTotal}% de daño.</Text>
      </View>

      <TouchableOpacity style={styles.attackButton} onPress={endFigth}>
        <Text style={styles.attackButtonText}>Reset</Text>
      </TouchableOpacity>
    </View>
  }

  return <View style={styles.container}>
    <Text style={styles.title}>
      Store of Potions
    </Text>
    {
      Boolean(potions.length) && (
        <View style={styles.gameStartContainer}>
          <Text style={styles.bagPotions}>{potions.length}</Text>
          <Text style={styles.bagPotions}>{potions.length === 1 ? "Potion" : "Potions"}</Text>
          <TouchableOpacity style={styles.attackButton} onPress={startFigth}>
            <Text style={styles.attackButtonText}>Attack</Text>
          </TouchableOpacity>
        </View>
      )
    }
    <View style={styles.row}>
      {
        [
          {
            alias: "true-love",
            name: "True love!",
            description: "Con esta poción caerá rendido/a a tus pies.",
            color: "red",
            avatar: "https://raw.githubusercontent.com/yellyoshua/kata-idukay-react/master/src/icons/potions/love-potion.png"
          },
          {
            alias: "false-illusions",
            name: "False illusions",
            description: "Corromperás su alma.",
            color: "blue",
            avatar: "https://raw.githubusercontent.com/yellyoshua/kata-idukay-react/master/src/icons/potions/illusions-potion.png"
          },
          {
            alias: "fake-friends",
            name: "Fake friends",
            description: "Lastimarás su simpatía.",
            color: "green",
            avatar: "https://raw.githubusercontent.com/yellyoshua/kata-idukay-react/master/src/icons/potions/fake-friends-potion.png"
          },
          {
            alias: "blessed-ignorance",
            name: "Blessed ignorance",
            description: "Acabarás con su carrera.",
            color: "yellow",
            avatar: "https://raw.githubusercontent.com/yellyoshua/kata-idukay-react/master/src/icons/potions/ignorance-potion.png"
          },
          {
            alias: "defame",
            name: "Defame",
            description: "Acabarás con su orgullo.",
            color: "gray",
            avatar: "https://raw.githubusercontent.com/yellyoshua/kata-idukay-react/master/src/icons/potions/defame-potion.png"
          }
        ].map((potion: PotionProps, index: number) => (
          <PotionCard key={index} potion={potion} onPressed={addPotion} />
        ))
      }
    </View>
  </View>
}

export function PotionCard({ potion, onPressed }: { potion: PotionProps; onPressed?: (potion: PotionProps) => void }) {
  return <TouchableOpacity style={styles.potionCard} onPress={() => {
    if (onPressed) {
      onPressed(potion)
    }
  }}>
    <View style={[styles.underline, { backgroundColor: potion.color }]} />
    <Image style={styles.potionCardImage} resizeMode="cover" source={{ uri: potion.avatar }} />
    <Text style={styles.potionCardTitle}>{potion.name}</Text>
    <Text style={styles.potionCardDescription}>{potion.description}</Text>
  </TouchableOpacity>
}

const styles = StyleSheet.create({
  container: {
    justifyContent: 'center'
  },
  title: {
    fontSize: 20,
    fontWeight: 'bold',
    color: "#fff",
    textAlign: "center"
  },
  row: {
    flex: 1,
    justifyContent: "center",
    flexDirection: "row",
    flexWrap: "wrap"
  },
  potionCard: {
    width: 150,
    margin: 10,
    padding: 10,
    backgroundColor: "#33363e",
    borderRadius: 10,
    textAlign: "center"
  },
  underline: {
    width: '100%',
    height: 2,
    marginBottom: 10
  },
  potionCardImage: {
    width: 120,
    height: 120,
    marginHorizontal: "auto"
  },
  potionCardTitle: {
    fontSize: 18,
    color: "#fff",
    fontWeight: "bold",
    paddingVertical: 10
  },
  potionCardDescription: {
    fontSize: 15,
    color: "#fff",
    paddingVertical: 10
  },
  gameStartContainer: {
    flex: 1,
    height: 42,
    flexDirection: "row",
    justifyContent: 'space-around',
    marginVertical: 10,
    borderRadius: 10,
    backgroundColor: "white",
    alignItems: "center",
    padding: 10
  },
  bagPotions: {
    fontSize: 17,
    color: 'black',
    fontWeight: "bold"
  },
  attackButton: {
    backgroundColor: "#FF5722",
    borderRadius: 10,
    padding: 10
  },
  attackButtonText: {
    fontSize: 17,
    color: 'white',
    fontWeight: "bold"
  },
  gameStats: {
    marginVertical: 10
  },
  gameStatsAttacks: {
    fontSize: 17,
    color: 'white',
    marginVertical: 10
  },
  gameStatsTotal: {
    fontSize: 17,
    color: 'white',
    fontWeight: "bold"
  },

});