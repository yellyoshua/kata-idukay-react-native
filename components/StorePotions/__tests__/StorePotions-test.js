import * as React from 'react';
import { render, fireEvent } from '@testing-library/react-native';
import renderer from 'react-test-renderer';

import StorePotions from '../StorePotionsComponent';

describe("unit test store-potions component",() => {

  const player = {name: "test",alias: "test-dev", avatar: "/my-icon.jpg"};
  const potions= {
    red: "True love!",
    blue: "False illusions",
    green: "Fake friends",
    yellow: "Blessed ignorance",
    gray: "Defame"
  };

  it(`renders correctly`, () => {
    const resetPlayer = jest.fn();
    const store = renderer.create(<StorePotions player={player} resetPlayer={resetPlayer}/>).toJSON();

    expect(store).toMatchSnapshot();
  });

  it("Be rendered all potions",() => {
    const resetPlayer = jest.fn();
    const {getByText} = render(<StorePotions player={player} resetPlayer={resetPlayer}/>);
  
    expect(getByText(potions.red)).toBeTruthy();
    expect(getByText(potions.blue)).toBeTruthy();
    expect(getByText(potions.green)).toBeTruthy();
    expect(getByText(potions.yellow)).toBeTruthy();
    expect(getByText(potions.gray)).toBeTruthy();
  });

  it(`Case 1`, () => {
    const resetPlayer = jest.fn();
    const {getByText} = render(<StorePotions player={player} resetPlayer={resetPlayer}/>);
  
    fireEvent.press(getByText(potions.red));
    fireEvent.press(getByText(potions.red));
    fireEvent.press(getByText(potions.blue));
    fireEvent.press(getByText(potions.green));

    fireEvent.press(getByText("Attack"));

    expect(getByText("Ataque 1: usar 3  pociones  distintas  causan  un 10% de daño.")).toBeTruthy();
    expect(getByText("Ataque 2: usar una  poción  causa  un 3% de daño.")).toBeTruthy();
    expect(getByText(`Total: ${player.name} ha causado un 13% de daño.`)).toBeTruthy();
    
    fireEvent.press(getByText("Reset"));

    expect(resetPlayer).toBeCalledWith(null);
  });

  it(`Case 2`, () => {
    const resetPlayer = jest.fn();
    const {getByText} = render(<StorePotions player={player} resetPlayer={resetPlayer}/>);
  
    fireEvent.press(getByText(potions.red));
    fireEvent.press(getByText(potions.red));
    fireEvent.press(getByText(potions.blue));
    fireEvent.press(getByText(potions.blue));
    fireEvent.press(getByText(potions.green));
    fireEvent.press(getByText(potions.yellow));
    fireEvent.press(getByText(potions.gray));

    fireEvent.press(getByText("Attack"));

    expect(getByText("Ataque 1: usar 5  pociones  distintas  causan  un 25% de daño.")).toBeTruthy();
    expect(getByText("Ataque 2: usar una  poción  causa  un 3% de daño.")).toBeTruthy();
    expect(getByText("Ataque 3: usar una  poción  causa  un 3% de daño.")).toBeTruthy();
    expect(getByText(`Total: ${player.name} ha causado un 31% de daño.`)).toBeTruthy();
    
    fireEvent.press(getByText("Reset"));

    expect(resetPlayer).toBeCalledWith(null);
  });

  it(`Case 3`, () => {
    const resetPlayer = jest.fn();
    const {getByText} = render(<StorePotions player={player} resetPlayer={resetPlayer}/>);
  
    fireEvent.press(getByText(potions.red));
    fireEvent.press(getByText(potions.red));
    fireEvent.press(getByText(potions.blue));
    fireEvent.press(getByText(potions.blue));
    fireEvent.press(getByText(potions.green));
    fireEvent.press(getByText(potions.green));
    fireEvent.press(getByText(potions.yellow));
    fireEvent.press(getByText(potions.gray));

    fireEvent.press(getByText("Attack"));

    expect(getByText("Ataque 1: usar 5  pociones  distintas  causan  un 25% de daño.")).toBeTruthy();
    expect(getByText("Ataque 2: usar 2  pociones  distintas  causan  un 5% de daño.")).toBeTruthy();
    expect(getByText("Ataque 3: usar una  poción  causa  un 3% de daño.")).toBeTruthy();
    expect(getByText(`Total: ${player.name} ha causado un 33% de daño.`)).toBeTruthy();
    
    fireEvent.press(getByText("Reset"));

    expect(resetPlayer).toBeCalledWith(null);
  });
})
