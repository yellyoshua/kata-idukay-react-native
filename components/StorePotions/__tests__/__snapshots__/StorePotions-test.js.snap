// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`unit test store-potions component renders correctly 1`] = `
<View
  style={
    Object {
      "justifyContent": "center",
    }
  }
>
  <Text
    style={
      Object {
        "color": "#fff",
        "fontSize": 20,
        "fontWeight": "bold",
        "textAlign": "center",
      }
    }
  >
    Store of Potions
  </Text>
  <View
    style={
      Object {
        "flex": 1,
        "flexDirection": "row",
        "flexWrap": "wrap",
        "justifyContent": "center",
      }
    }
  >
    <View
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "backgroundColor": "#33363e",
          "borderRadius": 10,
          "margin": 10,
          "opacity": 1,
          "padding": 10,
          "textAlign": "center",
          "width": 150,
        }
      }
    >
      <View
        style={
          Array [
            Object {
              "height": 2,
              "marginBottom": 10,
              "width": "100%",
            },
            Object {
              "backgroundColor": "red",
            },
          ]
        }
      />
      <Image
        resizeMode="cover"
        source={
          Object {
            "uri": "https://raw.githubusercontent.com/yellyoshua/kata-idukay-react/master/src/icons/potions/love-potion.png",
          }
        }
        style={
          Object {
            "height": 120,
            "marginHorizontal": "auto",
            "width": 120,
          }
        }
      />
      <Text
        style={
          Object {
            "color": "#fff",
            "fontSize": 18,
            "fontWeight": "bold",
            "paddingVertical": 10,
          }
        }
      >
        True love!
      </Text>
      <Text
        style={
          Object {
            "color": "#fff",
            "fontSize": 15,
            "paddingVertical": 10,
          }
        }
      >
        Con esta poción caerá rendido/a a tus pies.
      </Text>
    </View>
    <View
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "backgroundColor": "#33363e",
          "borderRadius": 10,
          "margin": 10,
          "opacity": 1,
          "padding": 10,
          "textAlign": "center",
          "width": 150,
        }
      }
    >
      <View
        style={
          Array [
            Object {
              "height": 2,
              "marginBottom": 10,
              "width": "100%",
            },
            Object {
              "backgroundColor": "blue",
            },
          ]
        }
      />
      <Image
        resizeMode="cover"
        source={
          Object {
            "uri": "https://raw.githubusercontent.com/yellyoshua/kata-idukay-react/master/src/icons/potions/illusions-potion.png",
          }
        }
        style={
          Object {
            "height": 120,
            "marginHorizontal": "auto",
            "width": 120,
          }
        }
      />
      <Text
        style={
          Object {
            "color": "#fff",
            "fontSize": 18,
            "fontWeight": "bold",
            "paddingVertical": 10,
          }
        }
      >
        False illusions
      </Text>
      <Text
        style={
          Object {
            "color": "#fff",
            "fontSize": 15,
            "paddingVertical": 10,
          }
        }
      >
        Corromperás su alma.
      </Text>
    </View>
    <View
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "backgroundColor": "#33363e",
          "borderRadius": 10,
          "margin": 10,
          "opacity": 1,
          "padding": 10,
          "textAlign": "center",
          "width": 150,
        }
      }
    >
      <View
        style={
          Array [
            Object {
              "height": 2,
              "marginBottom": 10,
              "width": "100%",
            },
            Object {
              "backgroundColor": "green",
            },
          ]
        }
      />
      <Image
        resizeMode="cover"
        source={
          Object {
            "uri": "https://raw.githubusercontent.com/yellyoshua/kata-idukay-react/master/src/icons/potions/fake-friends-potion.png",
          }
        }
        style={
          Object {
            "height": 120,
            "marginHorizontal": "auto",
            "width": 120,
          }
        }
      />
      <Text
        style={
          Object {
            "color": "#fff",
            "fontSize": 18,
            "fontWeight": "bold",
            "paddingVertical": 10,
          }
        }
      >
        Fake friends
      </Text>
      <Text
        style={
          Object {
            "color": "#fff",
            "fontSize": 15,
            "paddingVertical": 10,
          }
        }
      >
        Lastimarás su simpatía.
      </Text>
    </View>
    <View
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "backgroundColor": "#33363e",
          "borderRadius": 10,
          "margin": 10,
          "opacity": 1,
          "padding": 10,
          "textAlign": "center",
          "width": 150,
        }
      }
    >
      <View
        style={
          Array [
            Object {
              "height": 2,
              "marginBottom": 10,
              "width": "100%",
            },
            Object {
              "backgroundColor": "yellow",
            },
          ]
        }
      />
      <Image
        resizeMode="cover"
        source={
          Object {
            "uri": "https://raw.githubusercontent.com/yellyoshua/kata-idukay-react/master/src/icons/potions/ignorance-potion.png",
          }
        }
        style={
          Object {
            "height": 120,
            "marginHorizontal": "auto",
            "width": 120,
          }
        }
      />
      <Text
        style={
          Object {
            "color": "#fff",
            "fontSize": 18,
            "fontWeight": "bold",
            "paddingVertical": 10,
          }
        }
      >
        Blessed ignorance
      </Text>
      <Text
        style={
          Object {
            "color": "#fff",
            "fontSize": 15,
            "paddingVertical": 10,
          }
        }
      >
        Acabarás con su carrera.
      </Text>
    </View>
    <View
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "backgroundColor": "#33363e",
          "borderRadius": 10,
          "margin": 10,
          "opacity": 1,
          "padding": 10,
          "textAlign": "center",
          "width": 150,
        }
      }
    >
      <View
        style={
          Array [
            Object {
              "height": 2,
              "marginBottom": 10,
              "width": "100%",
            },
            Object {
              "backgroundColor": "gray",
            },
          ]
        }
      />
      <Image
        resizeMode="cover"
        source={
          Object {
            "uri": "https://raw.githubusercontent.com/yellyoshua/kata-idukay-react/master/src/icons/potions/defame-potion.png",
          }
        }
        style={
          Object {
            "height": 120,
            "marginHorizontal": "auto",
            "width": 120,
          }
        }
      />
      <Text
        style={
          Object {
            "color": "#fff",
            "fontSize": 18,
            "fontWeight": "bold",
            "paddingVertical": 10,
          }
        }
      >
        Defame
      </Text>
      <Text
        style={
          Object {
            "color": "#fff",
            "fontSize": 15,
            "paddingVertical": 10,
          }
        }
      >
        Acabarás con su orgullo.
      </Text>
    </View>
  </View>
</View>
`;
